relayer/workers/parachaincommitmentrelayer/beefy-listener.go

```diff
@@ -116,56 +137,54 @@ func (li *BeefyListener) subBeefyJustifications(ctx context.Context) error {
-			"relayChainBlockNumber": relayChainBlockNumber,
-			"ethereumBlockNumber":   event.Raw.BlockNumber,
-			"ethereumTxHash":        event.Raw.TxHash.Hex(),
+			"beefyBlockNumber":    beefyBlockNumber,
+			"ethereumBlockNumber": event.Raw.BlockNumber,
+			"ethereumTxHash":      event.Raw.TxHash.Hex(),
 		}).Info("Witnessed a new MMRRoot event")
 
-		li.log.WithField("blockNumber", relayChainBlockNumber).Info("Getting hash for relay chain block")
-		blockHash, err := li.relaychainConn.GetAPI().RPC.Chain.GetBlockHash(uint64(relayChainBlockNumber))
+		li.log.WithField("beefyBlockNumber", beefyBlockNumber).Info("Getting hash for relay chain block")
+		relayBlockHash, err := li.relaychainConn.GetAPI().RPC.Chain.GetBlockHash(uint64(beefyBlockNumber))
 		if err != nil {
 			li.log.WithError(err).Error("Failed to get block hash")
 			return err
 		}
-		li.log.WithField("blockHash", blockHash.Hex()).Info("Got blockhash")
-
-		// TODO this just queries the latest MMR leaf in the latest MMR and our latest parahead from the relaychain.
-		// we should ideally be querying the latest and last few leaves in the latest MMR until we find
-		// the first parachain block that has not yet been fully processed on ethereum,
-		// and then package and relay all newer heads/commitments together with their corresponding leaf
-		mmrProof := li.relaychainConn.GetMMRLeafForBlock(uint64(relayChainBlockNumber-1), blockHash)
-		allParaHeads, ourParaHead := li.GetAllParaheads(blockHash, OUR_PARACHAIN_ID)
+		li.log.WithField("relayBlockHash", relayBlockHash.Hex()).Info("Got relay chain blockhash")
 
-		ourParaHeadProof := createParachainHeaderProof(allParaHeads, ourParaHead)
-
-		messagePackets, err := li.extractCommitments(ourParaHead, mmrProof, ourParaHeadProof)
+		verifiedParaBlockNumber, verifiedParaBlockHash, err := li.relaychainConn.FetchLatestFinalizedParaBlock(
+			relayBlockHash, OUR_PARACHAIN_ID)
 		if err != nil {
-			li.log.WithError(err).Error("Failed to extract commitment and messages")
 			return err
 		}
-		if len(messagePackets) == 0 {
-			li.log.Info("Parachain header has no commitment with messages, skipping...")
-			continue
-		}
-		for _, messagePacket := range messagePackets {
-			li.log.WithFields(logrus.Fields{
-				"channelID":        messagePacket.channelID,
-				"commitmentHash":   messagePacket.commitmentHash,
-				"commitmentData":   messagePacket.commitmentData,
-				"ourParaHeadProof": messagePacket.paraHeadProof,
-				"mmrProof":         messagePacket.mmrProof,
-			}).Info("Beefy Listener emitted new message packet")
-
-			li.messages <- messagePacket
+
+		messagePackages, err := li.buildMissedMessagePackages(ctx, beefyBlockNumber, verifiedParaBlockNumber, verifiedParaBlockHash)
+		if err != nil {
+			li.log.WithError(err).Error("Failed to build missed message packages")
+			return err
 		}
 
+		li.EmitMessagePackages(messagePackages)
+
 	}
 	return nil
 }
 
+func (li *BeefyListener) EmitMessagePackages(packages []MessagePackage) {
```

Comment:
This func is private to BeefyListener. Having it public makes maintenance and refactoring a bit more difficult.

```suggestion
func (li *BeefyListener) emitMessagePackages(packages []MessagePackage) {
```

Reply:
https://github.com/Snowfork/polkadot-ethereum/pull/446/commits/b8c04ab57edecc49453d0f5791ca84b5a1343de0